crates/turbo-tasks-bytes/src/stream.rs

```diff
@@ -0,0 +1,200 @@
+use std::{
+    fmt,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context as TaskContext, Poll},
+};
+
+use anyhow::Result;
+use futures::{Stream as StreamTrait, StreamExt};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// Streams allow for streaming values from source to sink.
+///
+/// A Stream implements both a reader (which implements the Stream trait), and a
+/// writer (which can be sent to another thread). As new values are written, any
+/// pending readers will be woken up to receive the new value.
+#[derive(Clone, Debug)]
+pub struct Stream<T: Clone> {
+    inner: Arc<Mutex<StreamState<T>>>,
+}
+
+/// The StreamState actually holds the data of a Stream.
+struct StreamState<T> {
+    source: Option<Box<dyn StreamTrait<Item = T> + Send + Unpin>>,
+    pulled: Vec<T>,
+}
+
+impl<T: Clone> Stream<T> {
+    /// Constructs a new Stream, and immediately closes it with only the passed
+    /// values.
+    pub fn new_closed(pulled: Vec<T>) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(StreamState {
+                source: None,
+                pulled,
+            })),
+        }
+    }
+
+    /// Crates a new Stream, which will lazily pull from the source stream.
+    pub fn new_open<S: StreamTrait<Item = T> + Send + Unpin + 'static>(
+        pulled: Vec<T>,
+        source: S,
+    ) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(StreamState {
+                source: Some(Box::new(source)),
+                pulled,
+            })),
+        }
+    }
+
+    /// Returns a [StreamTrait] implementation to poll values out of our Stream.
+    pub fn read(&self) -> StreamRead<T> {
+        StreamRead {
+            source: self.clone(),
+            index: 0,
+        }
+    }
+
+    pub async fn into_single(&self) -> SingleValue<T> {
+        let mut stream = self.read();
+        let Some(first) = stream.next().await else {
+            return SingleValue::None;
+        };
+
+        if stream.next().await.is_some() {
+            return SingleValue::Multiple;
+        }
+
+        SingleValue::Single(first)
+    }
+}
+
+pub enum SingleValue<T> {
+    /// The Stream did not hold a value.
+    None,
+
+    /// The Stream held multiple values.
+    Multiple,
+
+    /// The held only a single value.
+    Single(T),
+}
+
+impl<T: Clone, S: StreamTrait<Item = T> + Send + Unpin + 'static> From<S> for Stream<T> {
+    fn from(source: S) -> Self {
+        Self::new_open(vec![], source)
+    }
+}
+
+impl<T: Clone> Default for Stream<T> {
+    fn default() -> Self {
+        Self::new_closed(vec![])
+    }
+}
+
+impl<T: Clone + PartialEq> PartialEq for Stream<T> {
+    // A Stream is equal if it's the same internal pointer, or both streams are
+    // closed with equivalent values.
```

Comment:
This has an interesting behavior where a === b can change after the streams have been read to the end.

Reply:
Also a === b might change after serialization...

Reply:
Do we need Eq?

Reply:
> This has an interesting behavior where a === b can change after the streams have been read to the end.

Yes, but it can only go from false -> true.

> Also a === b might change after serialization...

Not quite, it already needs to be closed before serialization, so it's the same "has the source been fully read" case.

> Do we need Eq?

Isn't that a requirement for use in a Vc?

Reply:
> Yes, but it can only go from false -> true.

hmm... technically this would be fine for Vcs, but it still feels a bit weird...

> Isn't that a requirement for use in a Vc?

You could opt-out of the eq check on celling...